src/porepy/grids/grid.py

```diff
@@ -1070,6 +1070,61 @@ def initiate_node_tags(self) -> None:
         values = [np.zeros(self.num_nodes, dtype=bool) for _ in keys]
         tags.add_tags(self, dict(zip(keys, values)))
 
+    def divergence(self, dim: Literal[1, 2, 3]) -> sps.csr_matrix:
+        """Get divergence operator for the grid.
+
+        If dim>=2, it is assumed that the first
+        column corresponds to the x-equation of face 0, second column is y-equation etc.
+        The next column is then the x-equation for face 1.
+        Correspondingly, the first row represents x-component in first cell etc.
+
+        Parameters:
+            dim: Dimension of the quantity of which we want to compute the divergence.
+
+        Raises:
+            ValueError: If ``dim`` is not 1, 2 or 3.
+
+        Returns:
+            Divergence operator. Dimensions: dim * (num_cells, num_faces)
+
+        """
+        if dim == 1:  # The divergence is a scalar
+            return self.cell_faces.T.tocsr()
+        elif dim in [2, 3]:  # The divergence is a vector
```

Comment:
Is the divergence a vector when dim==3? Or is it a matrix? And, is there a use case in PP which really allows for dim=3?

Reply:
Dim is the number of quantities per face, hence it is context dependent. A matrix (2-tensor) like stress (which would need a 3d representation, since we need to stay compatible with AdArrays, where the gradient adds a tensor dimension) must be unravelled to a vector before taking the divergence. This is unfortunate, but AFAIK, there are no good python libraries for sparse tensors of general order. 

See also answer elsewhere.